src/active-model/model.cr

```diff
@@ -416,127 +429,128 @@ abstract class ActiveModel::Model
-          {% else %}
-            {{name}}: { type: {{opts[:type_signature]}}, setter: false },
-          {% end %}
-        {% end %}
-      )
-
-      # :nodoc:
-      def initialize(%yaml : YAML::ParseContext, %node : ::YAML::Nodes::Node, _dummy : Nil, trusted = false)
-        previous_def(%yaml, %node, nil)
-        if !trusted
-          {% for name, opts in FIELDS %}
-            {% if !opts[:mass_assign] %}
-              @{{name}} = nil
-            {% end %}
-          {% end %}
-        end
-        apply_defaults
-        clear_changes_information
-      end
+    def self.from_json(string_or_io : String | IO, trusted : Bool = false) : self
+      super(string_or_io).tap &.after_initialize(trusted: trusted)
+    end
 
-      # Serialize from a trusted YAML source
-      def self.from_trusted_yaml(string_or_io : String | IO) : self
-        ctx = YAML::ParseContext.new
-        node = begin
-          document = YAML::Nodes.parse(string_or_io)
-
-          # If the document is empty we simulate an empty scalar with
-          # plain style, that parses to Nil
-          document.nodes.first? || begin
-            scalar = YAML::Nodes::Scalar.new("")
-            scalar.style = YAML::ScalarStyle::PLAIN
-            scalar
-          end
-        end
-        {{@type.name.id}}.new(ctx, node, nil, true)
-      end
+    # Deserializes the given JSON in *string_or_io* into
+    # an instance of `self`, assuming the JSON consists
+    # of an JSON object with key *root*, and whose value is
+    # the value to deserialize. Will not deserialise from
+    # fields with mass_assign: false
+    #
+    # ```
+    # class User < ActiveModel::Model
+    #   attribute name : String
+    #   attribute google_id : UUID, mass_assign: false
+    # end
+    #
+    # User.from_json(%({"main": {"name": "Jason", "google_id": "f6f70bfb-c882-446d-8758-7ce47db39620"}}), root: "main") # => #<User:0x103131b20 @name="Jason">
+    # ```
+    def self.from_json(string_or_io : String | IO, root : String, trusted : Bool = false) : self
+      super(string_or_io, root).tap &.after_initialize(trusted: trusted)
+    end
 
-      def assign_attributes_from_json(json)
-        json = json.read_string(json.read_remaining) if json.responds_to? :read_remaining && json.responds_to? :read_string
-        model = self.class.from_json(json)
-        data = JSON.parse(json).as_h
-        {% for name, opts in FIELDS %}
-          {% if opts[:mass_assign] %}
-            self.{{name}} = model.{{name}} if data.has_key?({{name.stringify}}) && self.{{name}} != model.{{name}}
-          {% end %}
-        {% end %}
+    # Serialize from a trusted JSON source
+    def self.from_trusted_json(string_or_io : String | IO) : self
+      self.from_json(string_or_io, trusted: true)
+    end
 
-        self
-      end
+    def self.from_trusted_json(string_or_io : String | IO, root : String, trusted : Bool = false) : self
```

Comment:
What's the reasoning for this and the alt on line 468? This creates a very ambiguous/potentially confusing API.

Reply:
That looks like a mistake to me tbh. 

@dukeraphaelng what do you think?

Reply:
https://github.com/spider-gazelle/active-model/blob/e874f25454226b5ad933670c2f32190e8216875e/src/active-model/model.cr#L440-L469

`self.from_json` and `self.from_trusted_json` can be specified with or without a json root.

This is consistent with the Crystal STDLIB API https://github.com/crystal-lang/crystal/blob/6d9a1d5830db5f276bf3df37fceeb0d5333e7e14/src/json/from_json.cr#L1-L29

Reply:
I think the issue is  the `trusted` argument